### PR TITLE
fix: fix client error 4391

### PR DESCRIPTION
**Pull request by samsja**
fixing https://github.com/jina-ai/jina/issues/4391

New behavior:
if an exception is raised then :

- [x] if there is an on_error callback then the callback is called and the original exception is passed to the on_error
- [x] Otherwise we just throw an ClientError which wrap the actual exception
- [x] docs

### CODE REVIEW

File: tests/unit/clients/python/test_on_err.py [hunk 1]
```diff
@@ -43,3 +49,42 @@ def validate(x):
...
+
+    on_error = OnError()
+
+    Client(host='0.0.0.0', protocol=protocol, port=12345).post(
+        '/blah',
+        inputs=DocumentArray.empty(10),
+        on_error=on_error,
+    )
+
+    assert on_error.is_called
```
[the remark this answers is not in the document]
Comment (samsja) [reply tests/unit/clients/python/test_on_err.py:90]: was not None you mean ? If there are no `exception` then on_error won't be called

File: jina/clients/base/grpc.py [hunk 2]
```diff
@@ -68,22 +68,25 @@ async def _get_results(
...
-            elif (
-                my_code == grpc.StatusCode.UNKNOWN
-                and 'asyncio.exceptions.TimeoutError' in my_details
-            ):
-                raise BadClientInput(
-                    f'{msg}\n'
-                    'often the case is that you define/send a bad input iterator to jina, '
-                    'please double check your input iterator'
-                ) from rpc_ex
+            if on_error:
```
Comment (JoanFM) [inline jina/clients/base/grpc.py:71]: I would still suggest to print the `log errors`

**Review by JoanFM — CHANGES_REQUESTED** [review]
Please make sure to update `documentation` stating the change of behavior from `3.1`

File: jina/clients/helper.py [hunk 3]
```diff
@@ -81,3 +81,31 @@ def callback_exec(
...
+
+    >>> def on_error(response, e = None):
+    >>>      print(e)
+    >>>      print(response.docs[0])
+    >>>
+    >>> except aiohttp.ClientError as e:
+    >>> callback_exec_on_error(on_error, e, logger)
+    """
+
+    on_error_wrap = lambda resp: on_error(resp, exception)
```
Comment (JoanFM) [inline jina/clients/helper.py:109]: for better backward compatibility, I would suggest to inspect if `on_error` accepts 2 parameters, if only 1, pass only resp.

**Review by JoanFM — CHANGES_REQUESTED** [review]

File: jina/clients/helper.py [hunk 4]
```diff
@@ -81,3 +81,31 @@ def callback_exec(
...
+    :param logger: a logger instance
+
+    Example:
+
+    >>> def on_error(response, e = None):
+    >>>      print(e)
+    >>>      print(response.docs[0])
+    >>>
+    >>> except aiohttp.ClientError as e:
+    >>> callback_exec_on_error(on_error, e, logger)
```
Comment (nan-wang) [inline jina/clients/helper.py:106]: The example codes is not complete. Please make a good example or leave it blank.
```suggestion
   .. highlight:: yaml
   .. code-block:: yaml
         def on_error(response, e = None):
              print(e)
              print(response.docs[0])
  
          except aiohttp.ClientError as e:
         callback_exec_on_error(on_error, e, logger)
```

File: jina/clients/helper.py [hunk 5]
```diff
@@ -81,3 +81,31 @@ def callback_exec(
...
+    exception: Exception,
+    logger: JinaLogger,
+    response: Optional = None,
+) -> None:
+    """Execute the on_error callback with the response, Use when anerror outside the responde status was thrown.
+
+    :param response: the response
+    :param on_error: the on_error callback
+    :param exception: the exception with was thrown and led to the call of on_error
+    :param logger: a logger instance
```
Comment (nan-wang) [inline jina/clients/helper.py:97]: the order of the docstrings is wired

**Review by nan-wang — DISMISSED** [review]

File: jina/clients/helper.py (hunk 4 above)
Comment (samsja) [reply jina/clients/helper.py:106]: Do you mean that I should add the import etc ?

File: jina/clients/helper.py [hunk 6]
```diff
@@ -81,3 +83,29 @@ def callback_exec(
...
+
+    if len(signature(on_error).parameters) == 1:
+        on_error_wrap = on_error
+        warnings.warn(
+            "on_error callback taking only the response parameters is deprecated. Please add one parameter "
+            "to handle additional optional Exception as well",
+            DeprecationWarning,
+        )
+    else:
+        on_error_wrap = lambda resp: on_error(resp, exception)
```
Comment (JoanFM) [inline jina/clients/helper.py:109]: I would like this to be done only once? Can we wrap the `on_error` method right when the `Client.post` is done? Like this we will do this only once for all the requests and not once per request?

What I mean, override the `on_error` by the wrap` thing that adds the extra `kwargs`

**Review by JoanFM — CHANGES_REQUESTED** [review]

File: jina/clients/mixin.py [hunk 7]
```diff
@@ -167,3 +170,23 @@ async def post(
...
+
+    num_args = len(signature(on_error).parameters)
+    if num_args == 1:
+        warnings.warn(
+            "on_error callback taking only the response parameters is deprecated. Please add one parameter "
+            "to handle additional optional Exception as well",
+            DeprecationWarning,
+        )
+
+        @wraps(on_error)
```
Comment (JoanFM) [inline jina/clients/mixin.py:185]: I am not sure u need this extra decoration? you are already wrapping it by returning `_fn`

File: jina/excepts.py [hunk 8]
```diff
@@ -1,93 +1,97 @@
...
     """Raised when the docker version is incompatible"""
 
 
-class DaemonInvalidDockerfile(FileNotFoundError, BaseJinaExeception):
+class DaemonInvalidDockerfile(FileNotFoundError, BaseJinaException):
     """Raised when invalid dockerfile is passed to JinaD"""
 
 
-class NoContainerizedError(Exception, BaseJinaExeception):
+class NoContainerizedError(Exception, BaseJinaException):
```
Comment (JoanFM) [inline jina/excepts.py:96]: Are we sure? not sure if it is safe from multi inheritance p.o.v

File: jina/clients/mixin.py (hunk 7 above)
Comment (JoanFM) [reply jina/clients/mixin.py:185]: Ok I see, this way it is more transparent!

File: jina/excepts.py (hunk 8 above)
Comment (samsja) [reply jina/excepts.py:102]: I added a deprecation warning so the old one is still usable

**Review by JoanFM — APPROVED** [review]

File: docs/fundamentals/executor/executor-in-flow.md [hunk 9]
```diff
@@ -325,7 +325,7 @@ class MyExecutor(Executor):
 f = Flow().add(uses=MyExecutor)
 
 
-def print_why(resp):
+def print_why(resp, exception: None):
```
Comment (JoanFM) [inline docs/fundamentals/executor/executor-in-flow.md:328]: exception: None seems wrong

**Review by JoanFM — CHANGES_REQUESTED** [review]

**Review by JoanFM — APPROVED** [review]

**Pull request merged**